pkg/storage/cacher.go

```diff
@@ -307,21 +381,53 @@ func (c *Cacher) Codec() runtime.Codec {
 	return c.storage.Codec()
 }
 
+func (c *Cacher) indexValues(event *watchCacheEvent) ([]string, bool) {
+	if c.indexFunc == nil {
+		return nil, false
+	}
+	// TODO: Currently we assume that only one index exists, thus
+	// exactly 1 value will be returned.
+	matchValues := c.indexFunc(event.Object)
+	if event.PrevObject == nil {
+		return []string{matchValues[0].Value}, true
+	}
+	prevMatchValues := c.indexFunc(event.Object)
+	if matchValues[0].Value == prevMatchValues[0].Value {
+		return []string{matchValues[0].Value}, true
+	}
+	return []string{matchValues[0].Value, prevMatchValues[0].Value}, true
+}
+
 func (c *Cacher) processEvent(event watchCacheEvent) {
+	indexValues, supported := c.indexValues(&event)
+
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
+	// Iterate over "allWatchers" no matter what the indexing function is.
+	for _, watcher := range c.watchers.allWatchers {
 		watcher.add(event)
 	}
+	if supported {
```

Comment:
Yes - it still works. Because this is just optimization. We are calling Filter function for all objects that are potential candidate at the end anyway. So everything is working.